knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go

```diff
@@ -1,51 +1,310 @@
-		return fmt.Errorf("failed to apply ConsoleCLIDownload manifest: %w", err)
+		return fmt.Errorf("failed to apply kn ConsoleCLIDownload resources manifest: %w", err)
+	}
+
+	if err := checkDeployments(&manifest, instance, apiclient); err != nil {
+		return err
 	}
+
 	return nil
 }
 
-// Delete deletes ConsoleCLIDownload for kn CLI download links
-func Delete(instance *servingv1alpha1.KnativeServing, apiclient client.Client) error {
-	log.Info("Deleting ConsoleCLIDownload CR for kn")
-	manifest, err := mfc.NewManifest(manifestPath(), apiclient, mf.UseLogger(log.WithName("mf")))
+// Check for deployments of kn ConsoleCLIDownload
+// This function is copied from knativeserving_controller.go in serving-operator
+func checkDeployments(manifest *mf.Manifest, instance *servingv1alpha1.KnativeServing, api client.Client) error {
+	log.Info("Checking deployments")
+	for _, u := range manifest.Filter(mf.ByKind("Deployment"), mf.ByKind("Route")).Resources() {
+		switch u.GetKind() {
+		case "Deployment":
+			deployment := &appsv1.Deployment{}
+			err := api.Get(context.TODO(), client.ObjectKey{Namespace: u.GetNamespace(), Name: u.GetName()}, deployment)
+			if err != nil {
+				return err
+			}
+			for _, c := range deployment.Status.Conditions {
+				if c.Type == appsv1.DeploymentAvailable && c.Status != corev1.ConditionTrue {
+					return fmt.Errorf("Deployment %q/%q not ready", u.GetName(), u.GetNamespace())
+				}
+			}
+		case "Route":
+			route := &routev1.Route{}
+			if err := api.Get(context.TODO(), client.ObjectKey{Namespace: u.GetNamespace(), Name: u.GetName()}, route); err != nil {
+				return err
+			}
+			host := getCanonicalHost(route)
+			if host == "" {
+				return fmt.Errorf("Route %q/%q not ready yet", u.GetName(), u.GetNamespace())
+			}
+		}
+	}
+	return nil
+}
+
+// applyKnConsoleCLIDownload applies kn ConsoleCLIDownload by finding
+// kn download resource route URL and populating spec accordingly
+func applyKnConsoleCLIDownload(apiclient client.Client, instance *servingv1alpha1.KnativeServing) error {
+
+	log.Info("Installing kn ConsoleCLIDownload")
+	route := &routev1.Route{}
+	knConsole := &consolev1.ConsoleCLIDownload{}
+
+	// find the route first
+	if err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: instance.GetNamespace(), Name: knConsoleCLIDownloadDeployRoute}, route); err != nil {
+		return fmt.Errorf("failed to find kn ConsoleCLIDownload deployment route")
+	}
+
+	knRoute := getCanonicalHost(route)
+	knConsoleObj := populateKnConsoleCLIDownload(https(knRoute), instance)
+
+	// Check if kn ConsoleCLIDownload exists
+	err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: "", Name: knCLIDownload}, knConsole)
+	switch {
+	case apierrors.IsNotFound(err):
+		log.Info("Creating kn ConsoleCLIDownload")
+		if err := apiclient.Create(context.TODO(), knConsoleObj); err != nil {
+			return err
+		} else {
+			// handle upgrade scenario: In 1.6 we'd created kn ConsoleCLIDownload CO with
```

Comment:
Why are we changing the name? We should be able to just reconcile in the needed changes I think?

Reply:
Changed its name from `kn` to `kn-cli-downloads` (following how its done for `oc`).
However, I think there is repetition, no need to explicitly name `cli-downloads`  to ConsoleCLIDownloads CO. A specific name should suffice. I will revert name change and respective code to handle it.